src/Api/TransactionsClient.php

```diff
@@ -85,7 +85,15 @@ public function refund($transaction_id, array $data) {
       return $response->data;
     }
 
-    throw new \Exception($response->data);
+    // Throw a decline exception on error.
+    $message = sprintf(
+      'Could not process the refund action for the transaction "%s"',
+      $transaction_id
+    );
+    $this->logger->warning($message);
+    throw new HardDeclineException('
```

Comment:
Wouldn't we throw the HardDeclineException for just the create payment errors because we're trying to charge the card right? Wouldn't a PaymentGatewayException do instead?

Reply:
Ohkay, Would it be better to use InvalidResponseException?

Reply:
I'm not sure what the correct exception type would be. I was looking at the way Authnet does it, and they actually return a PaymentGatewayException:

```php
  /**
   * {@inheritdoc}
   */
  public function refundPayment(PaymentInterface $payment, Price $amount = NULL) {
    /** @var \Drupal\commerce_payment\Entity\PaymentMethod $payment_method */
    $payment_method = $payment->getPaymentMethod();
    $this->assertPaymentState($payment, ['completed', 'partially_refunded']);
    // If not specified, refund the entire amount.
    $amount = $amount ?: $payment->getAmount();
    $this->assertRefundAmount($payment, $amount);

    $request = new CreateTransactionRequest($this->authnetConfiguration, $this->httpClient);
    $transaction_request = new TransactionRequest([
      'transactionType' => TransactionRequest::REFUND,
      'amount' => $amount->getNumber(),
      'refTransId' => $payment->getRemoteId(),
    ]);

    // Add customer Billing information to satisfy AVS.
    /** @var \Drupal\address\AddressInterface $address */
    $address = $payment_method->getBillingProfile()->address->first();
    $bill_to = array_filter([
      // @todo how to allow customizing this.
      'firstName' => $address->getGivenName(),
      'lastName' => $address->getFamilyName(),
      'company' => $address->getOrganization(),
      'address' => substr($address->getAddressLine1() . ' ' . $address->getAddressLine2(), 0, 60),
      'country' => $address->getCountryCode(),
      'city' => $address->getLocality(),
      'state' => $address->getAdministrativeArea(),
      'zip' => $address->getPostalCode(),
      // @todo support adding phone and fax
    ]);
    $transaction_request->addDataType(new BillTo($bill_to));

    // Adding order information to the transaction.
    $order = $payment->getOrder();
    $transaction_request->addOrder(new OrderDataType([
      'invoiceNumber' => $order->getOrderNumber() ?: $order->id(),
    ]));
    $transaction_request->addPayment(new CreditCardDataType([
      'cardNumber' => $payment_method->card_number->value,
      'expirationDate' => str_pad($payment_method->card_exp_month->value, 2, '0', STR_PAD_LEFT) . str_pad($payment_method->card_exp_year->value, 2, '0', STR_PAD_LEFT),
    ]));
    $request->setTransactionRequest($transaction_request);
    $response = $request->execute();

    if ($response->getResultCode() != 'Ok') {
      $this->logResponse($response);
      $message = $response->getMessages()[0];
      throw new PaymentGatewayException($message->getText());
    }

    $old_refunded_amount = $payment->getRefundedAmount();
    $new_refunded_amount = $old_refunded_amount->add($amount);
    if ($new_refunded_amount->lessThan($payment->getAmount())) {
      $payment->setState('partially_refunded');
    }
    else {
      $payment->setState('refunded');
    }

    $payment->setRefundedAmount($new_refunded_amount);
    $payment->save();
  }
```

Reply:
Ohkay.  `InvalidResponseException` comment says

```
/**
 * Thrown when the payment gateway receives an invalid response.
 *
 * The API endpoint might be down, throwing an error, or returning a response
 * whose signature can't be verified.
 */
```

and `PaymentGatewayException` is the Base exception for all payment gateway errors.

Wouldn't InvalidResponseException would be best fit for us?

Reply:
Ah, I shall modify it to be PaymentGatewayException.

Reply:
@shabananavas I have modified it to be PaymentGatewayException

Reply:
For PaymentGatewayException it say's:

`Base exception for all payment gateway errors.`

As we're not checking for any specific errors, this is probably best.